test/recipes/82-test_tfo_cli.t

```diff
@@ -0,0 +1,100 @@
+
+use strict;
+use warnings;
+
+use IPC::Open3;
+use Socket;
+use OpenSSL::Test qw/:DEFAULT srctop_file/;
+use OpenSSL::Test::Utils;
+
+setup("test_tfo");
+
+plan skip_all => "test_tfo_cli needs sock enabled" if disabled("sock");
+plan skip_all => "test_tfo_cli needs tls enabled" if disabled("tls");
+
+plan tests => 6;
+
+my $shlib_wrap = srctop_file("util", "shlib_wrap.sh");
+my $apps_openssl = srctop_file("apps", "openssl");
+my $cert = srctop_file("apps", "server.pem");
+my $client_good = 0;
+my $server_good = 0;
+my $connect_good = 0;
+
+SKIP: {
+    skip "TFO not enabled", 3 if disabled("tfo");
+
+    # try to reserve a port/get a random port
+    socket(SOCKET,PF_INET,SOCK_STREAM,(getprotobyname('tcp'))[2]);
+    bind( SOCKET, pack_sockaddr_in(0, inet_aton("127.0.0.1")));
+    my $mysockaddr = getsockname(SOCKET);
+    my ($port, $myaddr) = sockaddr_in($mysockaddr);
+    printf STDERR "Port: $port\n";
+    close(SOCKET);
+
+    my $spid = open3(my $sin, my $sout, my $serr, $shlib_wrap, $apps_openssl, "s_server", "-4", "-port", $port, "-cert", $cert, "-www", "-tfo");
+    my $cpid = open3(my $cin, my $cout, my $cerr, $shlib_wrap, $apps_openssl, "s_client", "-4", "-connect", ":$port", "-tfo");
+
+    print $cin "GET /\r\n";
+
+    waitpid($cpid, 0);
+    kill 'KILL', $spid;
+
+    while (<$cout>) {
+        chomp;
+        $client_good = 1 if /^Connecting via TFO$/;
+        $connect_good = 1 if /^Content-type: text/;
+    }
+    print STDERR "Client not attempting TFO\n" if ! ok($client_good);
+
+    while (<$sout>) {
+        chomp;
+        $server_good = 1 if /^Listening for TFO$/;
+    }
+    print STDERR "Server not attempting TFO\n" if ! ok($server_good);
+
+    print STDERR "Connection failed\n" if ! ok($connect_good);
+}
+$client_good = 1;
+$server_good = 1;
+$connect_good = 0;
+
+{
+    # try to reserve a port/get a random port
+    socket(SOCKET,PF_INET,SOCK_STREAM,(getprotobyname('tcp'))[2]);
+    bind( SOCKET, pack_sockaddr_in(0, inet_aton("127.0.0.1")));
+    my $mysockaddr = getsockname(SOCKET);
+    my ($port, $myaddr) = sockaddr_in($mysockaddr);
+    printf STDERR "Port: $port\n";
+    close(SOCKET);
+
+    my $spid = open3(my $sin, my $sout, my $serr, $shlib_wrap, $apps_openssl, "s_server", "-4", "-port", $port, "-cert", $cert, "-www");
+    my $cpid = open3(my $cin, my $cout, my $cerr, $shlib_wrap, $apps_openssl, "s_client", "-4", "-connect", ":$port");
+
+    print $cin "GET /\r\n";
+
+    waitpid($cpid, 0);
+    kill 'KILL', $spid;
+
+    while (<$cout>) {
+        chomp;
+        $client_good = 0 if /^Connecting via TFO$/;
+        $connect_good = 1 if /^Content-type: text/;
+    }
+    print STDERR "Client attempting TFO\n" if ! ok($client_good);
+
+    while (<$sout>) {
+        chomp;
+        $server_good = 0 if /^Listening for TFO$/;
+    }
+    print STDERR "Server attempting TFO\n" if ! ok($server_good);
+
+    print STDERR "Connection failed\n" if ! ok($connect_good);
+}
```

Comment:
Looks like these two tests are essentially copies of each other with some minor differences. Probably there are opportunities to reduce code duplication here and have a common function for these two variants of the same thing.

Reply:
The problem with the connectivity test is that there's no guarantee that it will connect in the CI. The tests pass when built locally.

Reply:
Hmmm. We open sockets etc with the TLSProxy tests and these reliably pass - so someway or other it should be possible.

Reply:
I think I see it... let me try something...

Reply:
It was non-obvious that `s_client` would accept a `0` port.